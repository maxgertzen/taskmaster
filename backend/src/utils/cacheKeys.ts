export const CACHE_KEYS = {
  TASK: (userId: string, taskId: string) =>
    `cache:user:${userId}:task:${taskId}` as const,
  TASK_LIST: (userId: string, listId: string) =>
    `cache:user:${userId}:tasks:${listId}` as const,
  LIST: (userId: string, listId: string) =>
    `cache:user:${userId}:list:${listId}` as const,
  LISTS: (userId: string) => `cache:user:${userId}:lists` as const,
  SEARCH_RESULTS: (userId: string, query: string) =>
    `cache:user:${userId}:search:${query}` as const,
} as const;
