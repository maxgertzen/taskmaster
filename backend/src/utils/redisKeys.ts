export const REDIS_KEYS = {
  USER: (userId: string) => `user:${userId}` as const,
  TASK: (userId: string, taskId: string) =>
    `user:${userId}:task:${taskId}` as const,
  TASK_LIST: (userId: string, listId: string) =>
    `user:${userId}:tasks:${listId}` as const,
  TASKS_INDEX: "tasks_index" as const,
  LIST: (userId: string, listId: string) =>
    `user:${userId}:list:${listId}` as const,
  LISTS: (userId: string) => `user:${userId}:lists` as const,
} as const;
