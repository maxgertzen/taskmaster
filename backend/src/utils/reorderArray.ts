export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const arr = [...array];
  const [movedItem] = arr.splice(fromIndex, 1);
  arr.splice(toIndex, 0, movedItem);
  return arr;
};
