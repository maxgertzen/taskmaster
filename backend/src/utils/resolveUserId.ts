import { getAppContainer } from "@src/container";

export async function resolveUserId(
  auth0Id: string,
  email?: string,
  name?: string
): Promise<string> {
  const container = getAppContainer();
  const usersService = container.cradle.usersService;
  const user = await usersService.getOrCreateUser(auth0Id, email, name);

  if (!user) {
    throw new Error(`User with auth0Id ${auth0Id} not found`);
  }

  if ("_id" in user) {
    return user._id?.toString() ?? "";
  }

  return user.auth0Id;
}
