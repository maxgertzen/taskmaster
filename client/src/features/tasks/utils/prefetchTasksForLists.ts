import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
