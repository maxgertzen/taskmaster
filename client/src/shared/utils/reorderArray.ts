export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
