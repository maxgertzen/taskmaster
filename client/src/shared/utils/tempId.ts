export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
