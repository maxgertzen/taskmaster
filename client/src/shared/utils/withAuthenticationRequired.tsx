import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
