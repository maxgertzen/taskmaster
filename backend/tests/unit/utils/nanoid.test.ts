import { generateUniqueId } from "@src/utils/nanoid";
import { utilTestConfigs } from "@tests/data/utils";

const { mockResponses } = utilTestConfigs;

jest.mock("nanoid", () => ({
  nanoid: jest.fn(() => mockResponses.nanoid.default),
}));

describe("Nanoid", () => {
  describe("generateUniqueId", () => {
    it("should return a unique id using nanoid", async () => {
      const result = await generateUniqueId();
      expect(result).toBe(mockResponses.nanoid.default);
    });

    it("should use dynamic import to load nanoid", async () => {
      await generateUniqueId();
      const { nanoid } = await import("nanoid");
      expect(nanoid).toHaveBeenCalled();
    });

    it("should return different ids on subsequent calls", async () => {
      const { nanoid } = await import("nanoid");
      const mockNanoid = nanoid as jest.Mock;

      mockNanoid
        .mockImplementationOnce(() => mockResponses.nanoid.sequence[0])
        .mockImplementationOnce(() => mockResponses.nanoid.sequence[1]);

      const id1 = await generateUniqueId();
      const id2 = await generateUniqueId();

      expect(id1).toBe(mockResponses.nanoid.sequence[0]);
      expect(id2).toBe(mockResponses.nanoid.sequence[1]);
      expect(id1).not.toBe(id2);
    });
  });
});
