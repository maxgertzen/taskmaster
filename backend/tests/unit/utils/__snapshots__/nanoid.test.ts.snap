// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nanoid snapshot should match the expected structure snapshot 1`] = `
{
  "edgeCases": {
    "emptyStrings": {
      "auth0Id": "",
      "listId": "",
      "taskId": "",
      "userId": "",
    },
    "specialChars": {
      "auth0Id": "auth0|special!@#",
      "listId": "list#789$",
      "taskId": "task#456$",
      "userId": "user@123!",
    },
  },
  "errors": {
    "importError": "Failed to import nanoid",
    "serviceError": "Service error",
    "userNotFound": "User with auth0Id auth0|123456 not found",
  },
  "mockResponses": {
    "nanoid": {
      "default": "mocked-nanoid",
      "sequence": [
        "id1",
        "id2",
        "id3",
      ],
    },
    "user": {
      "withId": {
        "_id": "user123",
        "auth0Id": "auth0|123456",
      },
      "withUndefinedId": {
        "_id": undefined,
        "auth0Id": "auth0|123456",
      },
      "withoutId": {
        "auth0Id": "auth0|123456",
      },
    },
  },
  "validInputs": {
    "auth0Id": "auth0|123456",
    "email": "test@example.com",
    "listId": "list789",
    "name": "Test User",
    "query": "searchQuery",
    "taskId": "task456",
    "userId": "user123",
  },
}
`;
